scenario/scenario.go

```diff
@@ -11,465 +14,352 @@ import (
-	require.NoError(t, err)
-
-	callOpts := &bind.CallOpts{
-		From: makerAddress,
-	}
-	erc1155AssetData, err := devUtils.EncodeERC1155AssetData(
-		callOpts,
-		constants.GanacheDummyERC1155MintableAddress,
-		[]*big.Int{tokenID},
-		[]*big.Int{erc1155FungibleAmount},
-		[]byte{},
-	)
-	require.NoError(t, err)
-
-	// Create order
-	testOrder := &zeroex.Order{
-		ChainID:               big.NewInt(constants.TestChainID),
-		MakerAddress:          makerAddress,
-		TakerAddress:          constants.NullAddress,
-		SenderAddress:         constants.NullAddress,
-		FeeRecipientAddress:   common.HexToAddress("0xa258b39954cef5cb142fd567a46cddb31a670124"),
-		MakerAssetData:        erc1155AssetData,
-		MakerFeeAssetData:     constants.NullBytes,
-		TakerAssetData:        common.Hex2Bytes("f47261b0000000000000000000000000871dd7c2b4b25e1aa18728e9d5f2af4c4e431f5c"),
-		TakerFeeAssetData:     constants.NullBytes,
-		Salt:                  big.NewInt(int64(time.Now().Nanosecond())),
-		MakerFee:              big.NewInt(0),
-		TakerFee:              big.NewInt(0),
-		MakerAssetAmount:      big.NewInt(1),
-		TakerAssetAmount:      zrxAmount,
-		ExpirationTimeSeconds: big.NewInt(time.Now().Add(24 * time.Hour).Unix()),
-		ExchangeAddress:       ganacheAddresses.Exchange,
-	}
-
-	// Sign Order
-	signedTestOrder, err := zeroex.SignTestOrder(testOrder)
-	require.NoError(t, err, "could not sign order")
-
-	// Set up balances/allowances
-
-	// SET ERC1155 allowance
-	txn, err = erc1155Mintable.SetApprovalForAll(makerOpts, ganacheAddresses.ERC1155Proxy, true)
-	require.NoError(t, err)
-	waitTxnSuccessfullyMined(t, ethClient, txn)
-
-	// All 1 billion ZRX start in this address
-	zrxCoinbase := constants.GanacheAccount0
-	if takerAddress == zrxCoinbase {
-		t.Errorf("takerAddress cannot be set to the ZRX coinbase address (e.g., the address with the 1 billion ZRX at Genesis)")
-	}
 
-	zrx, err := wrappers.NewZRXToken(ganacheAddresses.ZRXToken, ethClient)
-	require.NoError(t, err)
+	for _, tokenAmount := range tokenAmounts {
+		uri := ""
+		txn, err := erc1155Mintable.CreateWithType(opts, tokenAmount.tokenID, uri)
+		require.NoError(t, err)
+		waitTxnSuccessfullyMined(t, txn)
 
-	// Transfer ZRX to takerAddress
-	opts := &bind.TransactOpts{
-		From:   zrxCoinbase,
-		Signer: GetTestSignerFn(zrxCoinbase),
+		txn, err = erc1155Mintable.MintFungible(opts, tokenAmount.tokenID, []common.Address{traderAddress}, []*big.Int{tokenAmount.amount})
+		require.NoError(t, err)
+		waitTxnSuccessfullyMined(t, txn)
 	}
-	txn, err = zrx.Transfer(opts, takerAddress, zrxAmount)
-	require.NoError(t, err)
-	waitTxnSuccessfullyMined(t, ethClient, txn)
 
-	// SET ZRX allowance
-	opts = &bind.TransactOpts{
-		From:   takerAddress,
-		Signer: GetTestSignerFn(takerAddress),
-	}
-	txn, err = zrx.Approve(opts, ganacheAddresses.ERC20Proxy, zrxAmount)
+	// Set ERC1155 allowance
+	// HACK(albrow): erc1155Mintable does not allow setting allowance per token id.
```

Comment:
The `HACK` part comes from the fact that this is a behavior mismatch compared to every other token type. We usually only set the balances and allowances that are required for the order and nothing more. In this case I think that is not possible.

Reply:
That makes sense. I think `NOTE` would also be appropriate because it is non-standard, but it also isn't actually fixable. My interpretation of `HACK` is generally that it is something that we _can_ fix (even if we never do). We really can't fix this.